lib/internal/modules/cjs/loader.js

```diff
@@ -1195,29 +1195,37 @@ let warnRequireESM = true;
 Module._extensions['.js'] = function(module, filename) {
   if (filename.endsWith('.js')) {
     const pkg = readPackageScope(filename);
+    // Function require shouldn't be used in ES modules.
     if (pkg && pkg.data && pkg.data.type === 'module') {
+      const err = new ERR_REQUIRE_ESM(filename);
       if (warnRequireESM) {
         const parentPath = module.parent && module.parent.filename;
         const basename = parentPath &&
             path.basename(filename) === path.basename(parentPath) ?
           filename : path.basename(filename);
-        process.emitWarning(
+        const warningMsg =
           'require() of ES modules is not supported.\nrequire() of ' +
           `${filename} ${parentPath ? `from ${module.parent.filename} ` : ''}` +
           'is an ES module file as it is a .js file whose nearest parent ' +
           'package.json contains "type": "module" which defines all .js ' +
           'files in that package scope as ES modules.\nInstead rename ' +
           `${basename} to end in .cjs, change the requiring code to use ` +
           'import(), or remove "type": "module" from ' +
-          `${path.resolve(pkg.path, 'package.json')}.`,
+          `${path.resolve(pkg.path, 'package.json')}.\n`;
+        process.emitWarning(
```

Comment:
I don't know, at least on v13.2.0, -according to documentation- ES modules still on experimental phase. Maybe wait few weeks more, on my opinion.

Reply:
The warning was created when there was no error message here originally. Previously it was a sort of test to see if this would cause issues, but as from the discussion that led to the issue the warning itself is prohibitive to the use cases.

We should remove the warning entirely, that's the issue here.

Reply:
So, i guess we won't need [this conditional](https://github.com/nodejs/node/blob/141a6e34eed05577edf43ad085a74f330d0559cb/lib/internal/modules/cjs/loader.js#L1199), and could remove `warnRequireESM`variable (if we remove the warning, sure).

Also, the current format of the warning is quite complex for [error registering function](https://github.com/nodejs/node/blob/141a6e34eed05577edf43ad085a74f330d0559cb/lib/internal/errors.js#L291), by the way, this is the unique place were this error is thrown, I think we could leave error message as is, and just concatenate the warning message (as my changes suggest).

Let me know what do you think.

Reply:
The approach seems fine. One simplification would be to just make the second argument to the error the optional extra warning message which does the concatenation. Error messages can be functions to handle more complex cases fine usually.

Reply:
Oh, yess, sorry, I did not read all registered errors, I think you suggest somethign [like this](https://github.com/nodejs/node/blob/141a6e34eed05577edf43ad085a74f330d0559cb/lib/internal/errors.js#L1163).

So, your suggestions are change current error message (just concatenate current warning), and remove warning logic and stuff... am I right?

Reply:
Yes that sounds right.